.github/workflows/build-and-test.yml

```diff
@@ -1,415 +1,540 @@
+    strategy:
+      fail-fast: false
+      max-parallel: 4
+      matrix:
+        os: [windows-latest, macos-latest]
+
+    steps:
+      - name: Check out repo
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # Fetch complete history for accurate versioning
+
+      - name: Set up python environment
+        uses: ./.github/actions/setup_env
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+          requirements-files: requirements.txt requirements-dev.txt requirements-gui.txt
+          requirements: ${{ env.DEFAULT_DEPENDENCIES }}
+
+      - name: Make install
+        run: make install
+
+      - name: Set up node
+        uses: actions/setup-node@v2
+        with:
+          node-version: 14.x
+
+      - name: Restore node_modules cache
+        uses: actions/cache@v2
+        with:
+          path: workbench/node_modules
+          key: ${{ runner.os }}-${{ hashFiles('workbench/yarn.lock') }}
+
+      - name: Install workbench dependencies
+        working-directory: workbench
+        run: |
+          yarn config set network-timeout 600000 -g
+          yarn install
+
+      - name: Run workbench tests
+        working-directory: workbench
+        env:
+          CI: true
+        run: yarn test
+
+  build-binaries:
+    name: Build binaries
+    needs: check-syntax-errors
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [macos-10.15, windows-latest]
+        include:  # test with a range of numpy and gdal versions
+          - os: macos-10.15
+            binary-make-command: mac_dmg
+            puppeteer-log: ~/Library/Logs/invest-workbench/
+            workspace-path: InVEST-failed-mac-workspace.tar
+            binary-extension: dmg
+          - os: windows-latest
+            binary-make-command: windows_installer
+            puppeteer-log: ~/AppData/Roaming/invest-workbench/logs/
+            workspace-path: ${{ github.workspace }}
+            binary-extension: exe
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0 # Fetch complete history for accurate versioning
+
+        # Not caching chocolatey packages because the cache may not be reliable
+        # https://github.com/chocolatey/choco/issues/2134
+        # and this step only takes about 30 seconds.
+      - name: Install build dependencies (Windows)
+        if: matrix.os == 'windows-latest'
+        shell: powershell
+        run: |
+          choco install zip unzip
+          $env:PATH += ";C:\ProgramData\chocolatey\bin"
+          refreshenv  # Choco-provided command to reload environment variables
+
+      - name: Restore cached NSIS plugins (Windows)
+        id: restore-nsis-cache
+        if: matrix.os == 'windows-latest'
+        uses: actions/cache@v2
+        with:
+          path: |
+            C:\Program Files (x86)\NSIS\Include
+            C:\Program Files (x86)\NSIS\Plugins
+          key: NSIS-plugins

```

Comment:
What would you think about perhaps using a timed key here to ensure that the cache is refreshed periodically?

My main concern is that because we're using the github-managed NSIS, that restoring our cache will override any _other_ plugins and include scripts included in the default install if the NSIS version changes.  I'm pretty sure I remember there being other plugins and include headers in these directories.

Reply:
Oh good point. What if instead we just cache and restore each plugins and script individually? Then we won't overwrite any others. It'll probably make the cache a little smaller too.

Reply:
>What if instead we just cache and restore each plugins and script individually? Then we won't overwrite any others. It'll probably make the cache a little smaller too.

That would work!  But I think we'll still want the cache's key to be a hash of the NSIS install script.  That way if there are changes to the script, we'll get a cache miss and re-run the script.

The only thing about a more specific listing of files to cache is that we'd be introducing some repetition between the workflow and `ci/windows-ci-binary-install.ps1` (which is really now just an NSIS install script).  So if we add/remove a file to the `.ps1` script, we'll need to remember to also add it to the workflow, and since we don't change those very often, I could see us forgetting about that.  Then again, how often will we be adding an NSIS plugin?

Although a single source of truth might be ideal, it also sounds impractical. Maybe a big ol' comment in both places would be sufficient for us to remember to update the cache?  What do you think?

Side note, I'm super excited about caching helping us to avoid the unpredictability of sourceforge's servers.

Reply:
That's true, I doubt the NSIS script will need to change before we deprecate it with the Qt UI, but it would be best to hash it. I added `hashFiles('ci/windows-ci-binary-install.ps1')` to the key and added comments.